<?xml version="1.0"?>
<!--This is suppose to be a Well docomunted .launch file so we could take it as granted when working with other launch files.

A few notes:
    1. When working with traj path's, be sure to set the correct path.
    2.


-->

<launch>

    <!--    If planning to work with trajectories, each traj' should become an name (e.g "traj1")
            and than use them as list, in the desired drone.    -->
    <arg name="traj1" default="/home/galbra/catkin_ws/src/crazyflie_ros/crazyflie_mapping/scripts/takeoff.csv"/>
    <arg name="traj2" default="/home/galbra/catkin_ws/src/crazyflie_ros/crazyflie_mapping/scripts/figure8.csv"/>

    <!--    ***************************************************** Drones initiation ********************************************************
            Parameters:
                - uri:                              the drone hardware ID
                - frame:                            the ROS name space, for publishing and subscribing.
                - frame_active:                     flag, use this drone if activated.
                - frame__listening_to_key_board:    flag, for listening to key_board_teleop_twist node (ROS).
                - frame1_start_delay:               counter in sec, for delaying this drone's start.
                - traj_list:                        list of trajectoryis for this drone
                - initX, initY, initZ:              are relative to WORLD 0,0 in meters
    -->

    <!-- for first CF -->
    <arg name="uri1" default="radio://0/120/2M/E7E7E7E701"/>
    <arg name="frame1" default="cf1"/>
    <arg name="frame1_active" default="1"/>
    <arg name="frame1__listening_to_key_board" default="1"/>
    <arg name="frame1_start_delay" default="0"/>
    <arg name="traj_list" default="[$(arg traj1),$(arg traj2)]"/>
    <arg name="initX1" default="0.2"/><!--in meters-->
    <arg name="initY1" default="0.5"/> <!--in meters-->
    <arg name="initZ1" default="0"/>


    <!-- for second CF    -->
    <arg name="uri2" default="radio://0/120/2M/E7E7E7E707"/>
    <arg name="frame2" default="cf7"/>
    <arg name="frame2_active" default="1"/>
    <arg name="frame2__listening_to_key_board" default="1"/>
    <arg name="frame2_start_delay" default="14"/>
    <arg name="traj_list2" default="[$(arg traj1),$(arg traj2)]"/>
    <arg name="initX2" default="0.2"/><!--in meters-->
    <arg name="initY2" default="2.0"/><!--in meters-->
    <arg name="initZ2" default="0"/>


    <!-- for third CF    -->
    <arg name="uri3" default="radio://0/120/2M/E7E7E7E702"/>
    <arg name="frame3" default="cf3"/>
    <arg name="frame3_active" default="0"/>
    <arg name="frame3__listening_to_key_board" default="0"/>
    <arg name="frame3_start_delay" default="15"/>
    <arg name="traj_list3" default="[$(arg traj1),$(arg traj2)]"/>
    <arg name="initX3" default="0.0762"/><!--black closer right point-->
    <arg name="initY3" default="2.0574"/><!--black closer right point-->
    <arg name="initZ3" default="0"/>


    <!-- for forth CF    -->
    <arg name="uri4" default="radio://0/120/2M/E7E7E7E709"/>
    <arg name="frame4" default="cf9"/>
    <arg name="frame4_active" default="0"/>
    <arg name="frame4__listening_to_key_board" default="0"/>
    <arg name="frame4_start_delay" default="15"/>
    <arg name="traj_list4" default="[$(arg traj1),$(arg traj2)]"/>
    <arg name="initX4" default="0.0"/>
    <arg name="initY4" default="2.4"/>
    <arg name="initZ4" default="0"/>

    <!--    ***************************************************** End Drones initiation ********************************************************  -->

    <!--     ***************************************************** Arguments for indoor mapping  ********************************************************
        Parameters:
                - outer maze params (cm):           x_min, x_max,y_min ,y_max

                - env_lim:                          list of all maze params

                - resolution (int):                 "tail" size for grid calculations, and matplotlib plotting

                - nDrones (int):                    number of drones participating in mapping.
                #todo: make a counter in .launch file for nDrones

                - useRefEnv:                        temp development, should be removed.
    -->


    <arg name="x_min" value="0"/>
    <arg name="x_max" default="400"/><!-- in CENTIMETERS -->
    <arg name="y_min" default="0"/>
    <arg name="y_max" default="400"/><!-- in CENTIMETERS -->
    <arg name="env_lim" default="[$(arg x_min), $(arg x_max), $(arg y_min), $(arg y_max)]"/>
    <arg name="resolution" default="10"/> <!-- in CENTIMETERS per cell -->
    <arg name="env_space" default="0"/>
    <arg name="nDrones" default="2"/>
    <arg name="useRefEnv" default="0"/>
    <!--         ***************************************************** END Arguments for indoor mapping  ********************************************************   -->

    <!--    run rviz -->
    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find crazyflie_mapping)/launch/indoors.rviz"
          required="true"/>

    <!--    run bag record -->
<!--    <node pkg="rosbag" type="record" name="rosbag_record_standard_data" respawn="true" output="screen"-->
<!--          args="-o /home/$USER/ -a"/>-->

    <!--    Teleop_twist_keyboard  -->
    <node name="teleop_twist_keyboard" pkg="teleop_twist_keyboard" type="teleop_twist_keyboard.py" output="screen"
          launch-prefix="xterm -e"/>


    <!--    ***************************************************** VRPN NODE:  *****************************************************
            VRPN node is for using the OPTITRACK system.

            Params:
                - server: Need to be the same as "Local Interface" in MOTIVE streaming settings machine. and "Multicast Interface" should be set to this laptop IP
                - port (usually const): 3883
    -->
    <!--            <arg name="server" default="169.254.148.86"/>-->
    <!--            <arg name="port" default="3883"/>-->

    <!--        <node pkg="vrpn_client_ros" name="vrpn" output="screen" type="vrpn_client_node" required="true">-->
    <!--            <rosparam subst_value="true">-->
    <!--                server: $(arg server)-->
    <!--                port: $(arg port)-->
    <!--                update_frequency: 100.0-->
    <!--                frame_id: world-->
    <!--                use_server_time: false-->
    <!--                broadcast_tf: true-->
    <!--                refresh_tracker_frequency: 1.0 &lt;!&ndash; VRPN will add all trackers found on connection &ndash;&gt;-->
    <!--            </rosparam>-->
    <!--        </node>-->
    <!--    ***************************************************** END VRPN NODE:  *****************************************************     -->

    <!--    Activate the server.-->
    <include file="$(find crazyflie_driver)/launch/crazyflie_server.launch">
    </include>


    <!--    Set the name space of the CF_x to be its name-->
    <!--    ***************************************************** FRAME 1  ********************************************************-->
    <group ns="$(arg frame1)" if="$(eval arg('frame1_active') ==1)">
        <node pkg="crazyflie_driver" type="crazyflie_add" name="crazyflie_add" output="screen">
            <param name="uri" value="$(arg uri1)"/>
            <param name="tf_prefix" value="$(arg frame1)"/>

            <!--            Logging params:-->
            <param name="enable_logging" value="True"/>
            <param name="enable_logging_temperature" value="False"/>
            <param name="enable_logging_magnetic_field" value="False"/>
            <param name="enable_logging_pressure" value="False"/>
            <param name="enable_logging_imu" value="True"/>
            <param name="enable_logging_battery" value="False"/>

            <rosparam>
                genericLogTopics: ["log_pos" , "log_rpy" , "log_ranges", "log_sensors" ]
                genericLogTopicFrequencies: [20 , 20 , 20 , 20 ]
                genericLogTopic_log_pos_Variables: ["stateEstimate.x", "stateEstimate.y", "stateEstimate.z"]
                genericLogTopic_log_rpy_Variables: ["stabilizer.roll" , "stabilizer.pitch" , "stabilizer.yaw"]
                genericLogTopic_log_ranges_Variables: ["range.front", "range.back", "range.up", "range.left",
                "range.right" , "range.zrange" ]
                genericLogTopic_log_sensors_Variables: ["motion.deltaX" , "motion.deltaY" , "motion.outlierCount" ]
            </rosparam>
        </node>

        <!--        Tf_broadcast a static transform with offset between a drone launch point and world-->
        <node pkg="tf" type="static_transform_publisher" name="world_to_$(arg frame1)_takeoff"
              args="$(arg initX1) $(arg initY1) 0 0 0 0 1 world $(arg frame1)_takeoff 100"/>
        <!--        Tf_broadcast a specific drone with world-->
        <node name="tf_broadcast" pkg="crazyflie_mapping" type="tf_broadcaster.py" output="screen">
            <param name="tf_prefix" value="$(arg frame1)"/>
            <param name="initX" value="$(arg initX1)"/>
            <param name="initY" value="$(arg initY1)"/>
            <param name="initZ" value="$(arg initZ1)"/>
        </node>


        <!--        Publish TOF sensors readings as pointcloud-->
        <node name="points_publisher" pkg="crazyflie_mapping" type="publish_scan.py" output="screen">
            <param name="tf_prefix" value="$(arg frame1)"/>
        </node>




    </group>
    <!--    ***************************************************** END FRAME 1  ******************************************************** -->

       <!--    ***************************************************** FRAME 2  ********************************************************-->
    <group ns="$(arg frame2)" if="$(eval arg('frame2_active') ==1)">
        <node pkg="crazyflie_driver" type="crazyflie_add" name="crazyflie_add" output="screen">
            <param name="uri" value="$(arg uri2)"/>
            <param name="tf_prefix" value="$(arg frame2)"/>

            <!--            Logging params:-->
            <param name="enable_logging" value="True"/>
            <param name="enable_logging_temperature" value="False"/>
            <param name="enable_logging_magnetic_field" value="False"/>
            <param name="enable_logging_pressure" value="False"/>
            <param name="enable_logging_imu" value="True"/>
            <param name="enable_logging_battery" value="False"/>

            <rosparam>
                genericLogTopics: ["log_pos" , "log_rpy" , "log_ranges", "log_sensors" ]
                genericLogTopicFrequencies: [20 , 20 , 20 , 20 ]
                genericLogTopic_log_pos_Variables: ["stateEstimate.x", "stateEstimate.y", "stateEstimate.z"]
                genericLogTopic_log_rpy_Variables: ["stabilizer.roll" , "stabilizer.pitch" , "stabilizer.yaw"]
                genericLogTopic_log_ranges_Variables: ["range.front", "range.back", "range.up", "range.left",
                "range.right" , "range.zrange" ]
                genericLogTopic_log_sensors_Variables: ["motion.deltaX" , "motion.deltaY" , "motion.outlierCount" ]
            </rosparam>
        </node>




    <!--        Tf_broadcast a static transform with offset between a drone launch point and world-->
        <node pkg="tf" type="static_transform_publisher" name="world_to_$(arg frame2)_takeoff"
              args="$(arg initX2) $(arg initY2) 0 0 0 0 1 world $(arg frame2)_takeoff 100"/>
        <!--        Tf_broadcast a specific drone with world-->
        <node name="tf_broadcast" pkg="crazyflie_mapping" type="tf_broadcaster.py" output="screen">
            <param name="tf_prefix" value="$(arg frame2)"/>
            <param name="initX" value="$(arg initX2)"/>
            <param name="initY" value="$(arg initY2)"/>
            <param name="initZ" value="$(arg initZ2)"/>
        </node>


        <!--        Publish TOF sensors readings as pointcloud-->
        <node name="points_publisher" pkg="crazyflie_mapping" type="publish_scan.py" output="screen">
            <param name="tf_prefix" value="$(arg frame2)"/>
        </node>

    </group>
    <!--    ***************************************************** END FRAME 2  ******************************************************** -->


    <!--    ***************************************************** MAPPING modules *******************************************************-->
    <group ns="Mapping">
        <node name="grid_publisher" pkg="crazyflie_mapping" type="Grid.py" output="screen">
            <param name="env_lim" value="$(arg env_lim)"/>
            <param name="env_space" value="$(arg env_space)"/>
            <param name="resolution" value="$(arg resolution)"/>
            <param name="useRefEnv" value="$(arg useRefEnv)"/>
            <param name="excelPath" value="$(find crazyflie_mapping)/resources/MazeMap.csv"/>
            <param name="nDrones" value="$(arg nDrones)"/>
            <param name="drone_name_0" value="$(arg frame1)"/>
            <param name="drone_takeoff_position_0" value="[$(arg initX1), $(arg initY1), $(arg initZ1)]"/>
            <param name="drone_name_1" value="$(arg frame2)"/>
            <param name="drone_takeoff_position_1" value="[$(arg initX2), $(arg initY2), $(arg initZ2)]"/>
            <param name="drone_name_2" value="$(arg frame3)"/>
            <param name="drone_takeoff_position_2" value="[$(arg initX3), $(arg initY3), $(arg initZ3)]"/>

        </node>



    </group>

</launch>
